detection_rules/version_lock.py

```diff
@@ -0,0 +1,164 @@
+# 2.0.
+"""Helper utilities to manage the version lock."""
+from copy import deepcopy
+from typing import List, Optional
+
+import click
+
+from .rule import TOMLRule
+from .semver import Version
+from .utils import dict_hash, load_etc_dump, save_etc_dump, cached
+
+ETC_VERSION_LOCK_FILE = "version.lock.json"
+ETC_DEPRECATED_RULES_FILE = "deprecated_rules.json"
+MIN_LOCK_VERSION_DEFAULT = Version("7.13.0")
+
+
+def _convert_lock_version(stack_version: Optional[str]) -> Version:
+    """Convert an optional stack version to the minimum for the lock."""
+    if stack_version is None:
+        return MIN_LOCK_VERSION_DEFAULT
+    return max(Version(stack_version), MIN_LOCK_VERSION_DEFAULT)
+
+
+@cached
+def get_locked_version(rule_id: str, min_stack_version: Optional[str] = None) -> Optional[int]:
+    rules_versions = load_versions()
+
+    if rule_id in rules_versions:
+        latest_version_info = rules_versions[rule_id]
+        stack_version_info = latest_version_info.get("previous", {}).get(min_stack_version, latest_version_info)
+        return stack_version_info['version']
+
+
+@cached
+def get_locked_hash(rule_id: str, min_stack_version: Optional[str] = None) -> Optional[str]:
+    rules_versions = load_versions()
+
+    # Get the version info matching the min_stack_version if present
+    if rule_id in rules_versions:
+        latest_version_info = rules_versions[rule_id]
+        stack_version_info = latest_version_info.get("previous", {}).get(min_stack_version, latest_version_info)
+        existing_sha256: str = stack_version_info['sha256']
+        return existing_sha256
+
+
+def manage_versions(rules: List[TOMLRule],
+                    exclude_version_update=False, save_changes=False,
+                    verbose=True) -> (List[str], List[str], List[str]):
+    """Update the contents of the version.lock file and optionally save changes."""
+    from .packaging import current_stack_version
+
+    current_versions = deepcopy(load_versions())
+    versions_hash = dict_hash(current_versions)
+    rule_deprecations = load_etc_dump(ETC_DEPRECATED_RULES_FILE)
+
+    echo = click.echo if verbose else (lambda x: None)
+
+    already_deprecated = set(rule_deprecations)
+    deprecated_rules = set(rule.id for rule in rules if rule.contents.metadata.maturity == "deprecated")
+    new_rules = set(rule.id for rule in rules if rule.contents.latest_version is None) - deprecated_rules
+    changed_rules = set(rule.id for rule in rules if rule.contents.is_dirty) - deprecated_rules
+
+    # manage deprecated rules
+    newly_deprecated = deprecated_rules - already_deprecated
+
+    if not (new_rules or changed_rules or newly_deprecated):
+        return list(changed_rules), list(new_rules), list(newly_deprecated)
+
+    echo('Rule changes detected!')
+
+    for rule in rules:
+        if rule.contents.metadata.maturity == "production" or rule.id in newly_deprecated:
+            # assume that older stacks are always locked first
```

Comment:
how does this process look as a whole?

Reply:
like this (which I'll do in a GitHub action workflow in a PR shortly after this merges)
```console
$ git checkout 7.13
$ python -m detection_rules dev build-release -u

$ git checkout 7.14
$ python -m detection_rules dev build-release -u

$ git checkout 7.15
$ python -m detection_rules dev build-release -u
```